### ISSUE TITLE
Trigger gap-check toothDuration array samples time at any event?

### ISSUE THREAD

**Issue by 960** [rusefi/rusefi#1658]
There are a bug in trigger_decoder.cpp, so the gap-check doesnt do whats intended.

The array toothDurations[i] should only capture one edge, but instead it capture each event.

Ideally, it should look like this:

![msedge_XAP1FwEKRm](https://user-images.githubusercontent.com/9406140/88981946-67cc7980-d2c7-11ea-9ea7-d1e4584f713c.png)

But this is what it captures:

![stm32cubeide_rD64p2yqyA](https://user-images.githubusercontent.com/9406140/88981958-6dc25a80-d2c7-11ea-9ae8-79f04e31c990.png)

This means we get a array that looks like this at the gap:

With a VR-sensor, this means the tooth-durations will vary with rpm, and return errors.
(Falling edge in gap moves)
 
We also get very few actual teeth for the gap-check, so it works by luck.

```
#define getCurrentGapDuration(nowNt) \
	(isFirstEvent ? 0 : (nowNt) - toothed_previous_time)

	efitick_t currentDurationLong = getCurrentGapDuration(nowNt);

	toothDurations[0] = currentDurationLong > 10 * NT_PER_SECOND ? 10 * NT_PER_SECOND : currentDurationLong;


for (int i = GAP_TRACKING_LENGTH; i > 0; i--) {
			toothDurations[i] = toothDurations[i - 1];
		}
toothed_previous_time = nowNt;
```

**Comment by rusefillc** [comment]
Not an issue, code comments improved to reduce the chance of confusion.

**Comment by 960** [comment]
Not a issue?

toothed_previous_time are sampled at any event, and fills the array with both tooths and gaps.

I changed this to only sample at one edge, and see only the toothwidths + one long gap for missing tooth.
I also added correction for +-1 tooth.
That took care of my trigger issues.

Can you explain how/why it should be like it is?
Just wondering..

**Comment by rusefillc** [comment]
I believe that "toothed_previous_time are sampled at any event" is a FALSE statement

anything after the "toothed_previous_time are sampled at any event" FALSE statement is, well, not an issue.

See https://github.com/rusefi/rusefi/commit/74f8efd7e15778b511458297380c4ae99279711f commit which improved source code comments

**Comment by 960** [comment]
Ok, do however you want it.. 
I debugged while driving, and the strange tooth-times are caused by this edge moving a lot with VR.

Making one wide, and one narrow tooth..

Thats the two higher values you see in array

![stm32cubeide_rD64p2yqyA](https://user-images.githubusercontent.com/9406140/88986176-897f2e00-d2d2-11ea-8bec-2267a25f96fb.png)

**Comment by mck1117** [comment]
That's actually not your problem.  The problem is that these edges got missed:
![image](https://user-images.githubusercontent.com/568254/88987348-77799d00-d28a-11ea-918f-644a508c4fad.png)

The way that we count teeth has nothing to do with interrupts getting dropped, which is already noted at #1488.

For us to help you, you need to:
1. reproduce the issue using UNMODIFIED firmware & TS ini
2. share that msq

**Comment by 960** [comment]
Yeah, that was just a random image I drew at.

The case is the car doesnt stop anymore, so I dont notice the scheduler-thing.
It still trips the other codes, 6045 etc.. but after changing the gap-checking, at least it doesnt stop :-)

**Comment by mck1117** [comment]
Ok, so is there an issue or not? Does it reproduce on master or not?

**Comment by 960** [comment]
It's the same whatever firmware you use, just start debugging and see what values that goes in there.

Drama Queen spent only seconds to conclude this is not a issue and closed it.

I would call it a issue.
I think the correct durations to fill this array with are those shown in the bosch-screenshot.


These are the ones that ends up there

![Inkedpulseview_L08h7pwXKk_LI](https://user-images.githubusercontent.com/9406140/89032199-6edda100-d334-11ea-8308-b31755438a7c.jpg)

I changed it to capture times at only one edge(in trigger_central.cpp), and used these booleans instead.

![N5sKRcEDlc](https://user-images.githubusercontent.com/9406140/89032456-004d1300-d335-11ea-864b-07df6ec0d403.png)

**Comment by rusefillc** [comment]
insults are no acceptable, let me block you here.

**Comment by 960** [comment]
I would call it a insult to close this in seconds, and claim everything are right without even checking.

This is the same friggin story every time.

**Comment by 960** [comment]
Also, it syncs when toothDurations[0] are larger. 
This is wrong as well..

![stm32cubeide_oBjQywSbpG](https://user-images.githubusercontent.com/9406140/89052952-be34c900-d356-11ea-8404-350a49b1f93c.png)

**Comment by 960** [comment]
This is how the array looks in my code, so I dont need any help with this at all.
I tried to post a bug , that took you two seconds to rename to not beeing a bug and close.

I dont care about you at all Andrey, but I care about others that might get stuck or blow their engine over this.

Even if I think you are a sick creep, I would still like to see the project succeed.
When you go ahead with your closed-source plans maybe matt or someone else will take over and make it happen.

![stm32cubeide_ZwOe26T78x](https://user-images.githubusercontent.com/9406140/89079545-05d34900-d387-11ea-8c44-cdb73fafdcfc.png)

**Comment by mck1117** [comment]
Running code from master, I see tooth durations that are only looking at one edge, so the gaps are in the ratio 1-1-1-3-1 (etc).  What is the EXACT configuration you were using in your first post where you saw 1-2-2-1-1-1?

**Comment by mck1117** [comment]
Either way, whatever you did is just masking the real issue of missing teeth as identified in #1488.  That's still an issue, since we schedule all injection and ignition events off of trigger teeth.  If a tooth gets missed, everything for the rest of the cycle will end up late by the number of teeth missed (for example, 6 degrees for a 60-2 and one tooth lost).

**Comment by 960** [comment]
I am not able to get anything but the 1-2-2-1-1-1 ratio with any firmware or any settings.

The only way I get it right is to set these times in hwHandleShaftSignal.

In trigger_decoder.cpp, it gets those times no matter what..

**Comment by mck1117** [comment]
![image](https://user-images.githubusercontent.com/568254/89090156-e8d25200-d355-11ea-9c44-efe4e6dc902b.png)

**Comment by 960** [comment]
Whats been keeping mine from stopping are the change to the gap-check.

it always stopped when there was a delay too close to the gap, and firmware thougt it was the gap.

Now it doesnt "conclude" with seeing a gap until knowing what comes after the gap.

**Comment by 960** [comment]
I took one of the versions that didnt stop the engine, and the debugger shows the right ratios..
Tried a newer version, same trigger-folder and get the wrong ratios.

It cant figure where/how/why it gets those 1-2-2-1 duratons..

**Comment by mck1117** [comment]
have you touched this code at all? `trigger_central.cpp:291`:

```
	// for effective noise filtering, we need both signal edges, 
	// so we pass them to handleShaftSignal() and defer this test
	if (!CONFIG(useNoiselessTriggerDecoder)) {
		if (!isUsefulSignal(signal PASS_CONFIG_PARAMETER_SUFFIX)) {
			/**
			 * no need to process VR falls further
			 */
			return;
		}
	}
```

**Comment by 960** [comment]
No, I had the whole trigger-folder in sync all the time.

I moved the LogTriggerTooth(signal, timestamp PASS_ENGINE_PARAMETER_SUFFIX);

To after that function now, and this is how it get those times:

![javaw_oyJuaAmwo8](https://user-images.githubusercontent.com/9406140/89215218-846af900-d5c8-11ea-92ad-8467fc2d8c0f.png)

**Comment by 960** [comment]
// for effective noise filtering, we need both signal edges, 
	// so we pass them to handleShaftSignal() and defer this test
	if (signal == SHAFT_PRIMARY_RISING) {
			return;
		}
#if EFI_TOOTH_LOGGER
		LogTriggerTooth(signal, timestamp PASS_ENGINE_PARAMETER_SUFFIX);
#endif /* EFI_TOOTH_LOGGER */

If I "invert" that function, the times get right:

![javaw_r67N9FPZ9R](https://user-images.githubusercontent.com/9406140/89216840-6bb01280-d5cb-11ea-9856-340914278aa5.png)

**Comment by 960** [comment]
With VR, there will always be a signal in the gap.

So I guess this is all down to polarity?

I see bosch capture these times at falling edges..

**Comment by mck1117** [comment]
invert your trigger input polarity

**Comment by 960** [comment]
Where are the trigger-simulator signal feed?

**Comment by 960** [comment]
Inverting it looks better.
So correctly wired sensor = inverted?

Using the rising edge are not exactly correct.
Scoping with other ecu's shows opposite..

**Comment by 960** [comment]
Any way, this gap-check should be improved to avoid it seeing a gap at the first wider tooth it sees.
It even sees a gap with inverted VR and 2-2-1-1 ratio:

```
for (int i = 0;i<GAP_TRACKING_LENGTH;i++) {
				bool isGapCondition = cisnan(triggerShape->syncronizationRatioFrom[i]) || (toothDurations[i] > toothDurations[i + 1] * triggerShape->syncronizationRatioFrom[i]
					&& toothDurations[i] < toothDurations[i + 1] * triggerShape->syncronizationRatioTo[i]);
				isSync &= isGapCondition;
			}
```

**Comment by mck1117** [comment]
On 60-2, by default we look for a gap with ratio from 2.25 to 3.75.  That could probably be reduced.

**Comment by 960** [comment]
In VAG trigger it says 1.6 to 4.

Setting it to more than 2 and the simulator doesnt start.

Using rising edge for VR is just plain wrong.

The last tooth before gap will always be a falling edge

**Comment by 960** [comment]
With custom missing teeth trigger it doesnt start without inverting input..

The VAG 60-2 are totally wrong

**Comment by mck1117** [comment]
> With custom missing teeth trigger it doesnt start without inverting input..

correct, you need inverted input to be set.

**Comment by 960** [comment]
Tested with untouched trigger-folder, and with custom toothed wheel it died after around 500 m :-)

Changed to VAG-Decoder, and it stops at higher rpm's again..

Tooth logger stopping with vag-decoder

[2020-08-06_20.47.04.csv.zip](https://github.com/rusefi/rusefi/files/5037493/2020-08-06_20.47.04.csv.zip)

**Comment by mck1117** [comment]
> Tested with untouched trigger-folder, and with custom toothed wheel it died after around 500 m :-)

So it worked, is what I'm hearing.  Log of this happening? Logic analyzer?

**Comment by mck1117** [comment]
this is probably a duplicate of #988

**Comment by 960** [comment]
Ah, so it has been discussed before..

I have never seen the wheel with the wide tooth, but it's wrong to call it missing teeth :-)

At least the vag-trigger decoder is wrong, as it works inverted.
In tooth-logger as well it looks for two wider teeth either way.

Cant the trigger-stuff be put into classes as well?
Trigger_central, trigger_structure and trigger_decoder are pretty massive and confusing.

**Comment by 960** [comment]
With this tiny change, it seems to be running fine:

```
for (int i = 0;i<GAP_TRACKING_LENGTH;i++) {
		bool isGapCondition = cisnan(triggerShape->syncronizationRatioFrom[i]) || (toothDurations[i + 1] > toothDurations[i + 2] * triggerShape->syncronizationRatioFrom[i]
		&& toothDurations[i + 1] < toothDurations[i + 2] * triggerShape->syncronizationRatioTo[i] && toothDurations[i] * triggerShape->syncronizationRatioFrom[i] < toothDurations[i + 1]);
		isSync &= isGapCondition;
	}
```

**Comment by mck1117** [comment]
> I have never seen the wheel with the wide tooth, but it's wrong to call it missing teeth :-)

Volvo and BMW with flywheel VR sensors have a long tooth.  58 notches milled in to the flywheel, and a long part without a milled spot.

Either way, the two "flavors" of missing tooth produce identical signals, but inverted from one another if sensed with the same sensor.

**Comment by 960** [comment]
Ah, ok.. Never seen that one :-)

I have not had time to look any more at this lately.

Do you have any idea what can cause the delayed scheduler thing?


I think the whole trigger part is one huge mess, so could be cool to see how you would done it.. 
I still think you are a hell of a programmer Matt, and some of the issues you have solved before have been heavy :-)

**Comment by 960** [comment]
Almost lost interest for this stuff/all computer-stuff after holiday, but had a look at some of your branches..

You have actually done a lot cool stuff Matt :-)

Just had a drive with both your gaptracker and software-knock 👍 

Tested with real knock, but the gauge showed values between - 50 to - 30..

What kind of value should the variable knockBandCustom have?

**Comment by 960** [comment]
And can you please start to add your name as author in files?
In firmware, even if the whole file-content are replaced, they doesn't have your name.

It's useful to see who wrote the code..

**Comment by mck1117** [comment]
> And can you please start to add your name as author in files?

No, because git handles that. https://git-scm.com/docs/git-blame

**Comment by 960** [comment]
Ok, never seen that. I'll test it :-)

But that requires git? and are gone if this repo are deleted?
When working in eclipse or any editor you cant see it..

**Comment by 960** [comment]
Closed source at 500 sold ecus might sound familiar for some ;-)

**Comment by 960** [comment]
Your software-knock actually seems to work!!
How to add a second sensor?

**Comment by mck1117** [comment]
@960 can you share a datalog, tune, and the ignition table you used to induce knock?

**Comment by 960** [comment]
I'll try later today.

I have a KnockLite connected, and was just comparing the gauge with that one

**Comment by mck1117** [comment]
@960 have a chance to grab a log of knock occuring?

**Comment by 960** [comment]
I can try to get it done this weekend.

Have been focusing at the spi-fram at the spare time lately, and guess what..

It's working!! :-)

I never tought that I would be able, so this is a huge surprise..

What kind of values should I expect at the gauge?
(I am not sure which of the two pins that go to the connected sensor. One of them are connected to the knocklite)

**Comment by mck1117** [comment]
the exact values depend on your engine - what's important is the dynamic range between the conditions of "knock" and "not-knock".

It's a log scale, so -10 is actually 10x the intensity of -20.  I think with the dynamic range of our ADC, it should be able to sense from around -50 to +8.

**Comment by 960** [comment]
Matt are the only one who knows the firmware and behave adult, so a chat room I guess :-)

With the spi-fram I have a issue I cant figure..

It works correct first time after flashing when values are sent from TS.

But after reset, debugchannels doesnt work, and when I hit burn I get crc-error message.

I have updated the commands and tunerstudio.cpp

Do you see anything not normal here?

![stm32cubeide_y07nmWdxnf](https://user-images.githubusercontent.com/9406140/91646012-23113b00-ea4b-11ea-9e68-5d228c41d5da.png)

**Comment by 960** [comment]
Solved :-)

Now I can start testing the knock..

I am not going to generate any heavy knock, but I'll try to make it see what the knocklite does
(It has a led-scale)

**Comment by 960** [comment]
What's this in the knock-file?

palClearPad(GPIOD, 2);

**Comment by mck1117** [comment]
> What's this in the knock-file?
> 
> palClearPad(GPIOD, 2);

please run from master

that was testing logic that's now removed

**Comment by 960** [comment]
I just had a drive with the spi-fram and knock, but the ecu bricked after loading a tune..

Was sure it was the f-ram, so have been trying to reproduce the brick..

Just found the reason, and it was the software-knock triggering a hardfault..

Turned out this value was 600, even I have a maxvalue of 30 in TS..

Just added this:

if (CONFIG(knockBandCustom) > 30) {
	CONFIG(knockBandCustom) = 30;
	}

**Comment by 960** [comment]
Now it trips hardfault when it comes here..

addChannel("Baro Press", engineConfiguration->baroSensor.hwChannel, ADC_SLOW);

**Comment by 960** [comment]
Applied your fix in PR, and now it makes it to: 

 m_thread = chThdCreateStatic(m_threadstack, sizeof(m_threadstack), m_prio, StaticThreadTaskAdapter, this);

**Comment by 960** [comment]
Want some knock?

Made some "red" knock at the beginning, some normal driving then some "yellow" knock at the end

[2020-09-01_21.44.00.mlg.zip](https://github.com/rusefi/rusefi/files/5158502/2020-09-01_21.44.00.mlg.zip)

**Comment by 960** [comment]
![javaw_MxvoYf6wVT](https://user-images.githubusercontent.com/9406140/91902350-da06f400-eca1-11ea-9dc6-b23fc61a28d4.png)

Looks like this can actually be used for something with some tuning

**Comment by mck1117** [comment]
Ooh that does actually look like knock! "red" and "yellow" mean that you were comparing with your knocklite?

What did you have the knock frequency set to?

**Comment by 960** [comment]
Yes, its got a led-scale. It's very accurate compared to using headset.

 I think I had 12 in that setting.

The build had some other bugs, so I didnt get to test so much.
Will test some more when I got a better working firmware :-)

**Comment by mck1117** [comment]
cool - sounds good.

btw you do have per-cylinder info:
![image](https://user-images.githubusercontent.com/568254/91911425-ec633c80-ec65-11ea-81f7-6780cd801c37.png)

Might be an interesting test to add some advance/retard to one cylinder, and watch that one knock

**Comment by mck1117** [comment]
You might've actually had some knock on cylinder 2 earlier in the second pull:
![image](https://user-images.githubusercontent.com/568254/91917191-1ae81400-ec74-11ea-938a-05b97b8dbbc0.png)

**Comment by 960** [comment]
Yeah, but this is very sensitive..

The knocklink/knocklite detects knock much sooner than using headset.
They are calibrated to sense the engines natural noise first by starting calibration and running thru the rpm-range.

I guess thats the point it started flashing green :-)

I'll do some more testing when I have the firmware sorted again.
But it's clearly working.

Here's some interesting stoneage stuff from Autronic.

http://www.autronic.com.au/knock1_sound_rec_proc_22APR2016.pdf

**Comment by 960** [comment]
Looks like knock starts light from around -30 to heavy at around -15..

**Comment by 960** [comment]
How do I know what value the stacksize should be?

I am getting some errorcodes about lowstack in rpm_calculator sometimes when I hit burn

`
static THD_WORKING_AREA(eeThreadStack, 256);

class EeSpi: public eeSpiStream {
public:

	void send_offset(uint8_t cmd, uint32_t offset) const {
		b[0] = cmd;
		b[1] = (uint8_t)((offset >> 8) & 0xff);
		b[2] = (uint8_t)(offset & 0xff);
		spiSend(spid, 3, b);
	}

	bool read(uint32_t offset, size_t size, uint8_t *buf) {

		const uint8_t maxread = 128;
		while (size > maxread) {
			if (!read(offset, maxread, buf)) {
				return false;
			}
			offset += maxread;
			buf += maxread;
			size -= maxread;
		}

		for (uint8_t r = 0; r < FRAM_RETRIES; r++) {
			if (r != 0) {
				chThdSleepMilliseconds(FRAM_DELAY_MS);
			}
			{
				spiSelect(spid);
				send_offset(FRAM_CMD_READ, offset);
				spiReceive(spid, size, buf);
				spiUnselect(spid);
			}
			uint32_t crc1 = crc_crc32(0, buf, size);
			{
				spiSelect(spid);
				send_offset(FRAM_CMD_READ, offset);
				spiReceive(spid, size, buf);
				spiUnselect(spid);
			}
			uint32_t crc2 = crc_crc32(0, buf, size);

			if (crc1 == crc2) {
				// all good
				return true;
			}
		}

		return false;
	}


	bool write(uint32_t offset, size_t size, const uint8_t *buf) {

		const uint8_t maxread = 128;
		while (size > maxread) {
			if (!write(offset, maxread, buf)) {
				return false;
			}
			offset += maxread;
			buf += maxread;
			size -= maxread;
		}

		for (uint8_t r = 0; r < FRAM_RETRIES; r++) {
			if (r != 0) {
				chThdSleepMilliseconds(FRAM_DELAY_MS);
			}
			{
				tx[0] = FRAM_CMD_WREN;
				spiSelect(spid);
				spiSend(spid, 1, tx);
				spiUnselect(spid);
				spiSelect(spid);
				send_offset(FRAM_CMD_WRITE, offset);
				spiSend(spid, size, buf);
				spiUnselect(spid);
			}
			const uint8_t nverify = minI(size, sizeof(rbuf));
			uint32_t crc1 = crc_crc32(0, buf, nverify);

			spiSelect(spid);
			send_offset(FRAM_CMD_READ, offset);
			spiReceive(spid, nverify, rbuf);
			spiUnselect(spid);

			uint32_t crc2 = crc_crc32(0, rbuf, nverify);

			if (crc1 == crc2) {
				return true;
			}
		}

		return false;
	}
};
`

**Comment by 960** [comment]
And are there any point in splitting it in 128 buffers, or can I just transfer the 11K ?

Both works, but I dont know if that might overflow something..

**Comment by 960** [comment]
Did some tests with different frequency settings yesterday, but its only 12 that shows anything useful.

11 and 13 shows too much natural noise

**Comment by 960** [comment]
Finally a healthy firmware :-)

[2020-09-04_20.07.20.mlg.zip](https://github.com/rusefi/rusefi/files/5176558/2020-09-04_20.07.20.mlg.zip)

**Comment by mck1117** [comment]
I don't see any knock in that log

**Comment by 960** [comment]
Correct answer :-)

It's running at the very edge of knock-limit, with light knock at some areas.
For example at 40.838 there should be something.

I think you need to add some setup/autotune feature for it to be correct.

I have tried adjusting frequency around that area while idle or revving, and looked at what has given the lowest value.

For to tune it properly, I thing it needs a pull thru the rpm register without knock to set the frequency and level.

**Comment by 960** [comment]
New challenge :-)

Added 2 degrees more advance.

[2020-09-05_21.52.38.mlg.zip](https://github.com/rusefi/rusefi/files/5178555/2020-09-05_21.52.38.mlg.zip)

**Comment by 960** [comment]
I updated the trigger files, and now the trigger issues are back.

It's been running flawless for months when using gaptracker.h.

With the updates, I cant get gaptracker to work..

Can you fix it Matt?

I'll share a secret with you that took me days to figure out.
The third bit in composite-logger where you have TDC is actually a flag to toggle the top graph.
It's called "Trigger", and is set to 1 when cam is active.

<img width="959" alt="javaw_0vseX9TB0h" src="https://user-images.githubusercontent.com/9406140/95755764-1405e580-0ca5-11eb-8d0d-a98aaf31c23f.png">

**Comment by rusefillc** [comment]
Do we need gaptracker.h in master @mck1117 @960 ?

**Comment by 960** [comment]
I am not sure what it actually does compared to how it is.

But it's for sure the reason I have driven months without a misfire.
I have tried without it several times, and then the stopping engine issues returns..

Adding it back, and everything works fine.

**Comment by mck1117** [comment]
I *suspect* that "needing gaptracker.h" is a red herring.  I'd like to understand the actual problem before we start randomly changing code.

@960: You previously provided logic analyzer logs that showed trigger interrupts getting missed.  Does that still happen?

**Comment by 960** [comment]
I have not been troubleshooting this since then, as it's been running just fine..

everything are better than the engine stopping all the time when this is what I drive daily :-)

I think I have close to 100 000 Km on this soon..

**Comment by mck1117** [comment]
Nobody else has seen a consistent repro of this like you have, so I'd really appreciate your help so that we can improve the firmware and avoid this problem for you and others.  If you can try to repro on current firmware, I'm happy to add code to get more data about what what ECU is doing around the time it loses those teeth.

**Comment by 960** [comment]
It's been consistent between all kinds of updates and firmwares, and nothing seems to make any difference.

What does gaptracker actually do that is different to how it is now?

**Comment by mck1117** [comment]
I don't know that it is different or not.  I'm aware that nothing seems to make a difference.  That's why I want to collect more information so that we can conclusively say **WHY** the problem is happening, instead of shooting in the dark.  Can you get your setup back to a spot where it happens so we can work from there?

**Comment by 960** [comment]
It's no problem triggering the engine stopping, that happens every time I accelerate hard.

We do know what causes it to stop.
The delay causes gap-chech to see the gap at wrong place.

We can at least make sure that doesnt happen, and still investigate it.

Would it be difficult to move the check to one tooth after the gap?

**Comment by 960** [comment]
Now there are another user with the exact same issue.

I now have a couple of testers, and this one is in a Audi S3.

sparks was flying out his exhaust when it stalled, and he was scared the engine was blown up :-)

**Comment by rusefillc** [comment]
do you have final version of the hardware? can you ship me a couple of units? I am happy to buy a matching VAG in 2-3 months from now

**Comment by 960** [comment]
I wouldnt call it a final version :-)

Been looking for some real world testers for a while, so I first gave away some ecu's.

Turned out those targeted the wrong people (Broke kids wanting a ecu for free, now just waiting for someone to give them a free car to use it in).

The rest of the 5 old ones(with 8888) went in the trash.

I ordered 5 new ones, and sold those exept the one I use.
Now, finally they are in actual use :-)

I can order more after some validation.

Dont you think we should make the gap-check more secure? 
2-stage sync or something..

This happens when the scheduler are delayed and it sees the gap too early. (thats what we figured out I belive)
Each time it happens, it makes some real backfires and stops.
It sounds really unhealthy. One time, my car ran at only 3 cylinders for a while after, and I was sure I blew the engine..

**Comment by 960** [comment]
Trying to use the knock sensing do something, and made a knocklevel map with tresholds.

I have placed this inside getAdvance function, but not sure its expressed correct?^

getKnockLimit() are getting values from treshold table, and currentKnockLevel are db.

```
angle_t knockCorrection = 0;
		float currentKnockLevel = ENGINE(engineState.knockLevel);
		float knockLimit = getKnockLimit(rpm PASS_ENGINE_PARAMETER_SUFFIX);
		bool knock = (currentKnockLevel > knockLimit);

		if (knock) {
			for (knockCorrection = 0; knockCorrection < 8;  knockCorrection++) {
			if (!knock) {
				break;
			}
			}
		}
		angle -= knockCorrection;
```

**Comment by 960** [comment]
I can see it retards 8 degrees, but nothing between.
